[AMDGPU] Optionally Use GCNRPTrackers during scheduling

jrbyrnes opened the pull request:
This is part of a series of PRs which enable using the AMDGPU/GCN RPTrackers during scheduling. I've split them up to (hopefully) make reviewing easier. For context see https://github.com/llvm/llvm-project/pull/88797 . Since this is the final PR in the series, any high level comments should go here.

This PR adds the scheduling changes to: maintain the GCNRPTrackers during scheduling, and use them when making per-instruction scheduling decisions.

This PR is for commits beginning at [fbd0e54](https://github.com/llvm/llvm-project/pull/93090/commits/fbd0e54f749475c91a6fea60e84298cfdb043b38) and depends on https://github.com/llvm/llvm-project/pull/93088 and https://github.com/llvm/llvm-project/pull/93089

jrbyrnes committed:
[AMDGPU] Optionally Use AMDGPU RPTrackers during scheduling

arsenm commented on llvm/lib/Target/AMDGPU/GCNIterativeScheduler.cpp:
```diff
@@ -480,7 +480,7 @@ void GCNIterativeScheduler::scheduleLegacyMaxOccupancy(
                        "target occupancy = "
                     << TgtOcc << '\n');
-  GCNMaxOccupancySchedStrategy LStrgy(Context);
+  GCNMaxOccupancySchedStrategy LStrgy(Context, /*IsLegacyScheduler*/ true);
```
```suggestion
  GCNMaxOccupancySchedStrategy LStrgy(Context, /*IsLegacyScheduler=*/ true);
```

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1524,9 +1582,6 @@ bool PreRARematStage::sinkTriviallyRematInsts(const GCNSubtarget &ST,
     MachineInstr *OldMI = Entry.second;
 
-    // Remove OldMI from BBLiveInMap since we are sinking it from its MBB.
-    DAG.BBLiveInMap.erase(OldMI);
```
I'm not sure why you removed this line and added DAG.BBLiveInMap = DAG.getBBLiveInMap()  below

jrbyrnes replied:
Based on the way I'm reading the code and the comment, it seems the intention of this line is to remove any live-in registers that are defined by OldMI, as we are rematerailizing the registers. But, what this line actually does is remove the entire LiveIn LiveRegSet if OldMI is a key in the map.

It seems to me that we need to update the set of live-ins for the impacted region to reflect the eliminated cross-BB live ranges. Rather than doing this per-register, I've just recalculated the entire map after all the changes to LIS were made.

That said, perhaps this should be addressed separately?

vpykhtin replied:
That's a good idea to address that in a separate PR.

jrbyrnes replied:
Okay, I've removed it from this PR.

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.h:
```diff
@@ -70,6 +70,12 @@ class GCNSchedStrategy : public GenericScheduler {
   SmallVectorImpl<GCNSchedStageID>::iterator CurrentStage = nullptr;
 
+  // GCN RP Tracker for top-down scheduling
+  mutable GCNDownwardRPTracker TheTracker;
```
Maybe something like UpTracker and DownTracker for consistency?

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -658,6 +747,19 @@ void GCNScheduleDAGMILive::runSchedStages() {
+        GCNRPTracker::LiveRegSet *RegionLiveIns =
+            &LiveIns[Stage->getRegionIdx()];
+
+        reinterpret_cast<GCNRPTracker *>(DownwardTracker)
```
This looks criminal. You're forcing a call of a base class constructor for an object of derived type. Why this is needed?

vpykhtin replied:
Sorry I meant "reset" method, not constructor.  Basically you're leaving the derived object in undefined state.

jrbyrnes replied:
It was needed because the overloaded resets in GCNUpwardRPTracker/GCNDownwardRPTracker hide the GCNRPTracker::reset , fixed in https://github.com/llvm/llvm-project/pull/93088

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -128,23 +134,46 @@ void GCNSchedStrategy::initCandidate(SchedCandidate &Cand, SUnit *SU,
 
-  Pressure.clear();
-  MaxPressure.clear();
+  if (GCNTrackers) {
```
"else" from if at 138 line?

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -213,8 +242,15 @@ void GCNSchedStrategy::pickNodeFromQueue(SchedBoundary &Zone,
   if (DAG->isTrackingPressure()) {
-    SGPRPressure = Pressure[AMDGPU::RegisterPressureSets::SReg_32];
-    VGPRPressure = Pressure[AMDGPU::RegisterPressureSets::VGPR_32];
+    SGPRPressure =
```
This is hard to read. Can you do something like:

```
 if (!GCNTrackers) {
    // old code
 }  else {
    GCNRPTracker *T  = isTop? &DownwardTracker : &UpwardTracker;
    ... = T->getPressure().getVGPRNum(false);
    ... = T->getPressure().getSGPRNum();
 }
```

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -357,6 +393,16 @@ SUnit *GCNSchedStrategy::pickNode(bool &IsTopNode) {
+  if (GCNTrackers) {
+    MachineInstr *MI = SU->getInstr();
+    IsTopNode ? (void)DownwardTracker.advance(MI, false, DAG->getLIS())
+              : UpwardTracker.recede(*MI, false);
```
To me, if the node is scheduled then it should be "tracked"? If so, then the flag ShouldTrackIt isn't needed because the only use left is temp trackers where we shouldn't care?

jrbyrnes replied:
For the UpwardTracker, it is probably fine, but there will be issues with the DownwardTracker.

`ShouldTrackIt` is probably a misnomer and should be named to `UseInternalIterator`.

I don't think we can use the current version of the DownwardRPTracker for either use case. When scheduling a node, we have an issue with `advanceToNext`. It assumes that we can just advance the NextMI iterator to get the next instruction. This is only true if we have already scheduled the NextMI and updated the iterators, but we can't have scheduled the next instruction because we are in the process of scheduling the current instruction.

Do you see it differently?

vpykhtin replied:
Let's at least try to avoid complication of upward tracker interface.

jrbyrnes commented:
Address review comments + rebase on top of https://github.com/llvm/llvm-project/commit/d890dda16bf65bc36b783194afbe2ebc3e709afb

Pierre-vh commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -151,14 +157,42 @@ static bool canUsePressureDiffs(const SUnit &SU) {
+      TempTracker.getDownwardPressure(SU->getInstr(), Pressure, MaxPressure);
+    else
+      TempTracker.getUpwardPressure(SU->getInstr(), Pressure, MaxPressure);
+  } else {
```
early return instead of else

Pierre-vh commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -151,14 +157,42 @@ static bool canUsePressureDiffs(const SUnit &SU) {
+      auto MI = SU->getInstr();
+      TempTopTracker.advance(MI, false, DAG->getLIS());
+
+      Pressure[AMDGPU::RegisterPressureSets::SReg_32] =
```
Just assign a `GCNRegPressure` variable and put the assignments at the end, outside the branches?

jrbyrnes replied:
Hi thanks for taking a look.

I'm not sure about this one.

In order to sink the Pressure assignments without introducing a ternary, we would need to do something like:


```
  // GCNTrackers
  Pressure.resize(4, 0);
  GCNRPTracker TempTracker(
      AtTop ? static_cast<GCNRPTracker>(DownwardTracker)
            : static_cast<GCNRPTracker>(UpwardTracker));
  MachineInstr *MI = SU->getInstr();
  AtTop ? (void)reinterpret_cast<GCNDownwardRPTracker *>(&TempTracker)
              ->advance(MI, false, DAG->getLIS())
        : reinterpret_cast<GCNUpwardRPTracker *>(&TempTracker)
              ->recede(*MI);

  Pressure[AMDGPU::RegisterPressureSets::SReg_32] =
      TempTracker.getPressure().getSGPRNum();
  Pressure[AMDGPU::RegisterPressureSets::VGPR_32] =
      TempTracker.getPressure().getVGPRNum(false);
```


 I'm not sure I prefer this over what we have now. But if you feel strongly about it, I will change.

Pierre-vh commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -294,8 +330,16 @@ void GCNSchedStrategy::pickNodeFromQueue(SchedBoundary &Zone,
+      SGPRPressure = Pressure[AMDGPU::RegisterPressureSets::SReg_32];
+      VGPRPressure = Pressure[AMDGPU::RegisterPressureSets::VGPR_32];
+    } else {
+      GCNRPTracker *T = &UpwardTracker;
+      if (Zone.isTop())
+        T = &DownwardTracker;
```
```suggestion
      GCNRPTracker *T = Zone.isTop() ? &DownwardTracke : &UpwardTracker;
```

jrbyrnes replied:
Due to expression evaluation order, we'll get incompatible type error without the explicit casts

jrbyrnes commented:
Latest adds RP speculation by using the new interface in the GCNTrackers -- the review for this implementation, which has a more detailed git history for review purposes, is https://github.com/llvm/llvm-project/pull/93088 

Rebased for https://github.com/llvm/llvm-project/commit/825dbbbb94c985da4fd18da3e6e3baac05b11e23

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -288,6 +288,72 @@ collectVirtualRegUses(SmallVectorImpl<RegisterMaskPair> &RegMaskPairs,
+    }
+
+    return Result;
+  } else {
```
No else after return

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -288,6 +288,72 @@ collectVirtualRegUses(SmallVectorImpl<RegisterMaskPair> &RegMaskPairs,
+getLanesWithProperty(const LiveIntervals &LIS, const MachineRegisterInfo &MRI,
+                     bool TrackLaneMasks, Register RegUnit, SlotIndex Pos,
+                     LaneBitmask SafeDefault,
+                     bool (*Property)(const LiveRange &LR, SlotIndex Pos)) {
```
can this be function_ref

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -343,17 +409,47 @@ void GCNRPTracker::reset(const MachineInstr &MI,
+  MaxPressure = max(MaxPressure, CurPressure);
+  for (const RegisterMaskPair &P : DeadDefs) {
+    Register Reg = P.RegUnit;
+    if (!Reg.isVirtual())
```
dead defs increase pressure of physregs too?

jrbyrnes replied:
Currently, the functions used to maintain the current state of pressure / live regs (i.e. recede, and advance). So, if we try to track PhysRegs in our speculation queries, we will run into inconsistency issues. I think enabling PhysReg tracking should be a separate PR, perhaps required to turn on the trackers by default?

arsenm replied:
I currently have a patch stuck for over a year due to the pressure tracker brokenly handling dead defs

jrbyrnes replied:
Is this https://github.com/llvm/llvm-project/issues/76416 ?

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -288,6 +288,74 @@ collectVirtualRegUses(SmallVectorImpl<RegisterMaskPair> &RegMaskPairs,
+                                  const LiveIntervals *LIS,
+                                  bool Upward = false) {
+  for (const MachineOperand &MO : MRI.reg_nodbg_operands(Reg)) {
+    if (MO.isUndef())
+      continue;
```
No isUndef check, it is redundant with the readsReg case for uses and broken of the def of subregister case

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -414,6 +508,64 @@ void GCNUpwardRPTracker::recede(const MachineInstr &MI) {
   assert(CurPressure == getRegPressure(*MRI, LiveRegs));
 }
 
+void GCNUpwardRPTracker::bumpUpwardPressure(const MachineInstr *MI,
```
It's not really clear what this method does to the RP tracker object - modifies CurPressure and MaxPressure? What is the state of the object after the method call? From your code I only see one call against a temp object which may indicate that the state of a temp becomes undefined after a call. If so why don't make a non-method function returning max pressure?

jrbyrnes replied:
> It's not really clear what this method does to the RP tracker object - modifies CurPressure and MaxPressure? What is the state of the object after the method call? 

This method modifies CurPressure and MaxPressure w.r.t the uses/defs in MI. It is meant to be used for RP Speculation, so MI is not required to be the next instruction in the LIS. Due to this requirement, we can't use recede which uses the implied program order in LIS 
to calculate live lane masks.

> From your code I only see one call against a temp object which may indicate that the state of a temp becomes undefined after a call. If so why don't make a non-method function returning max pressure?

Since we are calculting the impact on RP for a SchedCandidate, we don't want to commit the new pressure state to the tracker (hence why we use the temp object). However, calculating RP w.r.t some instruction feels like the responsibility of the tracker, and keeping it as a method is consistent with the generic implementation.

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -414,6 +508,64 @@ void GCNUpwardRPTracker::recede(const MachineInstr &MI) {
+    CurPressure.inc(Reg, LiveAfter, LiveBefore, *MRI);
+  }
+  MaxPressure = max(MaxPressure, CurPressure);
+}
```
Overall algorithm is really cumbersome. Why do you need findUseBetween?

jrbyrnes replied:
> Why do you need findUseBetween?

For bumpUpwardPressure, we actually don't. Thanks for pointing this out. I used in an attempt to correct for a subreg liveness resulting in copying the generic implementation. This was caused by RegisterOperands::adjustLaneLiveness. This method sets the LaneMask of a use to the live lanes of the register at the position of the MI in the LIS. This adds lanes that are live at the MI, but not necessarily used by the MI. findUseBetween was meant to correct this, but this was not obvious and possibly incorrect. I've removed this and the call to and replaced adjustLaneLiveness with adjustDefLaneLiveness (for def lane handling).

arsenm commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -343,17 +407,47 @@ void GCNRPTracker::reset(const MachineInstr &MI,
   MaxPressure = CurPressure = getRegPressure(MRI_, LiveRegs_);
 }
 
+void GCNRPTracker::bumpDeadDefs(ArrayRef<RegisterMaskPair> DeadDefs) {
```
That is the name in the generic tracker, it's best to match the names here as a clone

jrbyrnes commented:
I spent a while looking RP speculation in the face of dead lanes in superreg defs. In the current code (as propagated from the generic trackers), we basically just ignore these lanes when calcuting pressure impacts for speculative RP. However, this is inaccurate. For example, if we have an instruction which defines a vreg_128 and but only subregs 1 and 2 are used, then bumpDownwardPressure will speculate that this vgpr def increase VGPR pressure by 2. Since those dead regs are not needed after the instruction, this RP change is consistent with RAs view of conflicts for positions after the instruction. However, during the instruction (before the dead slot), we temporarily increase VGPR pressure by 4 to provide registers for those dead lanes. However, this may mislead the scheduler, particularly for bottom up scheduling in which defs reduce RP. Encoding this into RPTracker's CurPressure doesn't seem to make much sense, as this is used to model the pressure after the instruction (whether top-down or bottom-up direction). So, I think we should probably introduce a separate GCNRegPressure field in the trackers to model any temporary increases to RP, and the scheduler can use this to more accurately set heuristics. That said, this should probably be addressed in separate PR.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -58,11 +58,17 @@ static cl::opt<bool>
                cl::init(false));
 
+static cl::opt<bool> GCNTrackers(
+    "amdgpu-use-amdgpu-trackers", cl::Hidden,
```
Needs some tests using this, where this shows a difference

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -414,6 +534,49 @@ void GCNUpwardRPTracker::recede(const MachineInstr &MI) {
+    LaneBitmask LiveBefore = (LiveAfter & ~DefLanes) | UseLanes;
+
+    CurPressure.inc(Reg, LiveAfter, LiveAfter & LiveBefore, *MRI);
+    MaxPressure = max(MaxPressure, CurPressure);
```
We do decrease pressure for defs (inc with PrevMask >= NewMask lowers to a decrease), but I've deleted the increase pressure for defs. 


```
  // Kill liveness at live defs.
  for (const RegisterMaskPair &P : RegOpers.Defs) {
    Register Reg = P.RegUnit;
    LaneBitmask LiveAfter = LiveRegs.contains(Reg);
    LaneBitmask UseLanes = getRegLanes(RegOpers.Uses, Reg);
    LaneBitmask DefLanes = P.LaneMask;
    LaneBitmask LiveBefore = (LiveAfter & ~DefLanes) | UseLanes;

    // There may be parts of the register that were dead before the
    // instruction, but became live afterwards. Similarly, some parts
    // may have been killed in this instruction.
    decreaseRegPressure(Reg, LiveAfter, LiveAfter & LiveBefore);
    increaseRegPressure(Reg, LiveAfter, ~LiveAfter & LiveBefore);
  }
  // Generate liveness for uses.
  for (const RegisterMaskPair &P : RegOpers.Uses) {
    Register Reg = P.RegUnit;
    // If this register was also in a def operand, we've handled it
    // with defs.
    if (getRegLanes(RegOpers.Defs, Reg).any())
      continue;
    LaneBitmask LiveAfter = LiveRegs.contains(Reg);
    LaneBitmask LiveBefore = LiveAfter | P.LaneMask;
    increaseRegPressure(Reg, LiveAfter, LiveBefore);
  }
}

```

The "missing" line is `increaseRegPressure(Reg, LiveAfter, ~LiveAfter & LiveBefore);` . This line exists in generic trackers to handle those registers which appear in both uses and defs of an instruction. The generic tracker then skips over these registers when considering the uses in the subsequent loop.

In the "missing" `increaseRegPressure`, the lane masks seem incompatible. The `NewMask` ( `~LiveAfter & LiveBefore` ) has no overlap with `LiveAfter`. For this reason alone, we can't 1:1 port it to our tracker. Additionally, this only works in the generic tracker because the `increaseRegPressure` exits if `NewMask` (which reduces to `~LiveAfter & UseLanes`) has no active lanes (which is often the case). We can't add this exit to our `inc` method as the `NewMask` can have no active lanes -- a common use case when we close out a range and want to decrement the pressure.

We could add this check to `bumpUpwardPressure` and/or fix the masks, but, to me, this "missing" line is a pretty confusing way to increase the RP for any new use lanes introduced by the instruction. In this PR, I've just handled these regs in the use loop.

jrbyrnes replied:
See also: https://github.com/llvm/llvm-project/pull/109875

jrbyrnes commented:
Added several tests based on the more significant lit changes.

Since this is a change which impacts scheduling heuristics, the changes are a mix of regressions and improvements. I've analyzed and included all the significant regressions, and added comments / fixmes to expplain the situation. The causes are: 1. not tracking PhysRegs, 2. the scheduler impoving RP in initial stage and not triggering high-RP stage, and 3. the scheduler being overly sensitive to minor changes in RP, and not being willing to incur a small penalty to enable scheduling of RP reducing instructions.

These will be addresed in a follow-up PR which enables the GCN RP Tracker features by default.

I've also included several tests where things go as expected -- more accurate RP Tracking leads to reductions in RP and, ultimately, improving occupancy

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNRegPressure.cpp:
```diff
@@ -430,28 +593,44 @@ bool GCNDownwardRPTracker::reset(const MachineInstr &MI,
+             ? CurrLIS->getInstructionIndex(*LastTrackedMI).getDeadSlot()
+             : CurrLIS->getInstructionIndex(*NextMI).getBaseIndex();
+  } else { //! UseInternalIterator
+    CurrLIS = TheLIS;
```
I'm not sure what is the CurrLIS? Are there more than one LISes anywhere?

vpykhtin commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -148,17 +154,44 @@ static bool canUsePressureDiffs(const SUnit &SU) {
+    Pressure[AMDGPU::RegisterPressureSets::AGPR_32] =
+        TempDownwardTracker.getPressure().getAGPRNum();
+  } else {
+    GCNUpwardRPTracker TempUpwardTracker(UpwardTracker);
```
So this is the cornerstone of you change. You're keeping two trackers consistent with the order of the scheduling and at every candidate you need to find its impact on the pressure. My concerns are below.

You're making a full copy of tracker object for every candidate at each step. This is not cheap as you're making a deep copy of DenseMap for live register set in every tracker just to throw it away immediately with no use. This would be ok for the start and would be ok if it made the program as simple as:
    temp_tracker(consistent_tracker)
    temp_tracker.advance/recede
    temp_tracker.getPressure
    get rid of temp_tracker
But you've created tons of code just to accomplish above. I believe the reason is that LIS isn't consistent with the incoming candidate.

bumpDownwardPressure it's a harder task with non-consistent LIS. Can you do it as a _const_ method of the GCNDownwardTracker? This means that it should not modify the state of the tracker in anyway and preferably with no extra data member to the object itself (except mutables if it's required for some caching). This way it wouldn't require DenseMap copy and make the intention if bumpDownwardPressure clear: It works on a tentative schedule with non-updated LIS.

Upward direction is different. GCNUpwardTracker is created in mind to be used in the situation when LIS isn't updated. The only place where LIS is used by the tracker is the determination of the live mask for register uses. It just reads the mask from the non-updated LIS and the rationale for this is that the defs should dominate uses for any valid schedule and the mask from the non-updated LIS should work. In other words GCNUpwardTracker  should be able to be used like the snippet above without additional work.

jrbyrnes replied:
Unfortunately, it is not this easy for the upwardRPTracker

The method the upwardTracker uses to determine the use lane masks adds lanemasks from other subranges unrelated to the the subreg use in the instruction.

On the other hand, bumpUpwardPressure uses RegOpers.collect which simply uses TRI.getSubRegIndexLaneMask(SubRegIdx) to get the lane masks.

In short, recede and bumpupward are logically inconsistent for this usecase, and I would prefer to not change recede. This would result in introducing behavioral changes when trackers are disabled.

jrbyrnes replied:
This is coming from a real example which has code like

```
%0:vreg_64 = IMPLICIT_DEF 
%1:vgpr_32 = IMPLICIT_DEF, implicit %0.sub0
%2:vgpr_32 = IMPLICIT_DEF, implicit %0.sub1
...
```

After bottomup scheduling the instruction defining `%2`, we'll speculate the RP for the instruction defining `%1`. When speculating using recede, we use `getLiveLaneMask` to calculate the mask of `%0.sub0`. To do so, we iterate over the `subranges()` of `%0` and find both `sub0` and `sub1` to be live at the position. Thus we inaccurately calculate the usemask of `%0.sub0` as 0xF (for both `sub0` and `sub1`). This is because `%0.sub1` is live at the index for the instruction defining `%1`. 

Whereas `getSubRegIndexLaneMask` (used by `bumpUpwardPressure`) for the `sub0` subregidx just returns 0xC as the mask.

For this small example, there will be no problem. But with wide regs with more subregs, these inaccuracies will effect scheduling. This doesn't impact RP calculations for whole regions with consistent `LIS` ordering because the use mask inaccuracies don't effect the delta between previous use mask and new use mask.

vpykhtin replied:
After (in bottom-up order) the instruction defining %2 we have %0:1100b live for %0.sub1. After we speculate over instruction defining %1, we get %0:1111b live indeed because now sub0 and sub1 are live. But when we calculate pressure we're accounting for a mask increase from %0:1100b to %0:1111b that is for a 0011b difference and this seems correct.

It seems though that it would be more correct if the mask found by getLiveLaneMask would be ANDed with the mask of the actually used subreg in the instruction. This way it can allow to find live masks for a complicated cases like sub1_sub0 subregs (it can have sub0, sub1 or sub1 and sub0 lives) but not wider than actually used subreg mask.

jrbyrnes replied:
Yes you are correct, there is no issue with this example. I simplified out the problem.

Please consider

```
---
name:            pdiff
tracksRegLiveness: true
body:             |
  bb.0:
  %49:areg_1024 = IMPLICIT_DEF
  undef %63.sub0:vreg_128 = COPY %49.sub0:areg_1024
  %63.sub1:vreg_128 = COPY %49.sub1:areg_1024
  %63.sub2:vreg_128 = COPY %49.sub2:areg_1024
  %63.sub3:vreg_128 = COPY %49.sub3:areg_1024
  
  S_ENDPGM 0, amdgpu_allvgprs
...
```


When speculating the first bottom-up instruction, we will calculate the RP from each of the COPYs. If we do so using recede, we will look at the live lanes of the register based on the LIS ordering (`collectVirtualRegUses` -> `getLiveLaneMask`). For the instruction defining undef %63.sub0, %49 has a LaneMask of 0xFF (all four lanes), The RPTracker would then determine this single lane copy uses 4 VGPRs. Similar problems for the instructions defining sub1 and sub2. Moreover, after scheduling an instruction bottomup, we will still have incorrect difference based RP calculations. bumpUpwardPressure calculates the usemask by doing something similar to what you suggest -- it will calculate the lanemask directly from the subreg being used by the instruction. As a result, it doesn't have these issues.

recede really doesn't work as is for this usecase. As you mentioned, we can adjust it to work for speculation, but I would prefer to contain the changes under the optional flag if possible.

jrbyrnes commented:
Force push to rebase + solve merge conflicts

jrbyrnes commented:
Fix recede to handle LIS program order violations + delete bumpUpwardPressure

jrbyrnes commented:
Latest passes psdb with GCNTrackers enabled

vpykhtin commented:
No more objections from my side given that the patch is under a CL flag.

jrbyrnes commented:
Having issues with CLI, force-push to land via web UI
